server/src/main/java/org/elasticsearch/index/seqno/RetentionLease.java

```diff
@@ -92,14 +92,17 @@ public RetentionLease(final String id, final long retainingSequenceNumber, final
         if (timestamp < 0) {
             throw new IllegalArgumentException("retention lease timestamp [" + timestamp + "] out of range");
         }
-        Objects.requireNonNull(source);
```

Comment:
Even though the following line would throw the same NPE, I think it is nice to keep this as explicit requirement (similar one to above requireNonNull for the id)